File: backend/routes/bugs.py
from fastapi import APIRouter, HTTPException, Depends
from typing import Optional
from datetime import datetime, timezone
from database import db, get_next_key
from auth_helpers import get_current_user
from models import BugInput, BugUpdate
import uuid

router = APIRouter(prefix="/api/bugs", tags=["bugs"])

@router.get("")
async def get_bugs(project_id: Optional[str] = None, assignee_id: Optional[str] = None, status: Optional[str] = None, priority: Optional[str] = None, user=Depends(get_current_user)):
    query = {}
    if project_id:
        query["project_id"] = project_id
    if assignee_id:
        query["assignee_id"] = assignee_id
    if status:
        query["status"] = status
    if priority:
        query["priority"] = priority
    return await db.bugs.find(query, {"_id": 0}).sort("created_at", -1).to_list(500)

@router.get("/all")
async def get_all_bugs(user=Depends(get_current_user)):
    bugs = await db.bugs.find({}, {"_id": 0}).sort("created_at", -1).to_list(500)
    for b in bugs:
        proj = await db.projects.find_one({"id": b["project_id"]}, {"_id": 0, "name": 1, "key": 1, "color": 1})
        b["project"] = proj
    return bugs

@router.post("")
async def create_bug(input: BugInput, project_id: str, user=Depends(get_current_user)):
    project = await db.projects.find_one({"id": project_id}, {"_id": 0})
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    key = await get_next_key(project_id, project["key"])
    doc = {
        "id": str(uuid.uuid4()), "project_id": project_id, "key": key,
        "title": input.title, "description": input.description,
        "status": input.status, "priority": input.priority,
        "assignee_id": input.assignee_id, "attachments": [],
        "created_by": user["_id"],
        "created_by_name": user.get("name", ""),
        "created_at": datetime.now(timezone.utc).isoformat(),
        "updated_at": datetime.now(timezone.utc).isoformat(),
        "comment_count": 0,
    }
    await db.bugs.insert_one(doc)
    doc.pop("_id", None)
    # Log "reported" activity
    await db.comments.insert_one({
        "id": str(uuid.uuid4()), "entity_id": doc["id"], "entity_type": "bug",
        "user_id": user["_id"], "user_name": user.get("name", ""),
        "content": "reported this bug", "type": "activity",
        "created_at": datetime.now(timezone.utc).isoformat(),
    })
    if input.assignee_id and input.assignee_id != user["_id"]:
        await db.notifications.insert_one({
            "id": str(uuid.uuid4()), "user_id": input.assignee_id,
            "type": "bug_assigned", "message": f"Bug {key}: {input.title} assigned to you",
            "link": f"/project/{project_id}/bugs", "read": False,
            "created_at": datetime.now(timezone.utc).isoformat(),
        })
    return doc

@router.put("/{bug_id}")
async def update_bug(bug_id: str, input: BugUpdate, user=Depends(get_current_user)):
    old = await db.bugs.find_one({"id": bug_id}, {"_id": 0})
    if not old:
        raise HTTPException(status_code=404, detail="Bug not found")
    update_data = {k: v for k, v in input.model_dump().items() if v is not None}
    update_data["updated_at"] = datetime.now(timezone.utc).isoformat()
    await db.bugs.update_one({"id": bug_id}, {"$set": update_data})
    # Log status change activity
    if "status" in update_data and update_data["status"] != old.get("status"):
        await db.comments.insert_one({
            "id": str(uuid.uuid4()), "entity_id": bug_id, "entity_type": "bug",
            "user_id": user["_id"], "user_name": user.get("name", ""),
            "content": f"changed status to {update_data['status'].replace('_', ' ').title()}",
            "type": "activity", "created_at": datetime.now(timezone.utc).isoformat(),
        })
        target = old.get("assignee_id") or old.get("created_by")
        if target and target != user["_id"]:
            await db.notifications.insert_one({
                "id": str(uuid.uuid4()), "user_id": target,
                "type": "status_changed", "message": f"{old['key']} status changed to {update_data['status']}",
                "link": f"/project/{old['project_id']}/bugs", "read": False,
                "created_at": datetime.now(timezone.utc).isoformat(),
            })
    # Log assignee change activity
    if "assignee_id" in update_data and update_data["assignee_id"] != old.get("assignee_id"):
        if update_data["assignee_id"]:
            assignee_user = await db.users.find_one({"_id": __import__('bson').ObjectId(update_data["assignee_id"])})
            assignee_name = assignee_user.get("name", "someone") if assignee_user else "someone"
            await db.comments.insert_one({
                "id": str(uuid.uuid4()), "entity_id": bug_id, "entity_type": "bug",
                "user_id": user["_id"], "user_name": user.get("name", ""),
                "content": f"assigned this to {assignee_name}",
                "type": "activity", "created_at": datetime.now(timezone.utc).isoformat(),
            })
    return await db.bugs.find_one({"id": bug_id}, {"_id": 0})

@router.delete("/{bug_id}")
async def delete_bug(bug_id: str, user=Depends(get_current_user)):
    await db.bugs.delete_one({"id": bug_id})
    await db.comments.delete_many({"entity_id": bug_id})
    return {"message": "Bug deleted"}

# ─── Attachments ───
from pydantic import BaseModel

class AttachmentPayload(BaseModel):
    file_name: str
    file_url: str
    file_type: str = "image/png"

@router.post("/{bug_id}/attachments")
async def add_attachment(bug_id: str, payload: AttachmentPayload, user=Depends(get_current_user)):
    bug = await db.bugs.find_one({"id": bug_id}, {"_id": 0})
    if not bug:
        raise HTTPException(status_code=404, detail="Bug not found")
    att = {
        "id": str(uuid.uuid4()),
        "file_name": payload.file_name,
        "file_url": payload.file_url,
        "file_type": payload.file_type,
        "uploaded_by": user["_id"],
        "created_at": datetime.now(timezone.utc).isoformat(),
    }
    await db.bugs.update_one({"id": bug_id}, {"$push": {"attachments": att}})
    return att

@router.delete("/{bug_id}/attachments/{attachment_id}")
async def remove_attachment(bug_id: str, attachment_id: str, user=Depends(get_current_user)):
    await db.bugs.update_one({"id": bug_id}, {"$pull": {"attachments": {"id": attachment_id}}})
    return {"message": "Attachment removed"}

File: backend/routes/bugs.js
const express = require('express');
const { v4: uuidv4 } = require('uuid');
const Bug = require('../models/Bug');
const Project = require('../models/Project');
const Comment = require('../models/Comment');
const Notification = require('../models/Notification');
const Counter = require('../models/Counter');
const User = require('../models/User');
const auth = require('../middleware/auth');

const router = express.Router();

// GET /api/bugs
router.get('/', auth, async (req, res) => {
  try {
    const query = {};
    if (req.query.project_id) query.project_id = req.query.project_id;
    if (req.query.assignee_id) query.assignee_id = req.query.assignee_id;
    if (req.query.status) query.status = req.query.status;
    if (req.query.priority) query.priority = req.query.priority;
    const bugs = await Bug.find(query).select('-_id -__v').sort({ created_at: -1 }).limit(500).lean();
    res.json(bugs);
  } catch (err) {
    res.status(500).json({ detail: err.message });
  }
});

// GET /api/bugs/all
router.get('/all', auth, async (req, res) => {
  try {
    const bugs = await Bug.find({}).select('-_id -__v').sort({ created_at: -1 }).limit(500).lean();
    for (const b of bugs) {
      const proj = await Project.findOne({ id: b.project_id }).select('name key color -_id').lean();
      b.project = proj;
    }
    res.json(bugs);
  } catch (err) {
    res.status(500).json({ detail: err.message });
  }
});

// POST /api/bugs?project_id=X
router.post('/', auth, async (req, res) => {
  try {
    const { project_id } = req.query;
    const project = await Project.findOne({ id: project_id }).lean();
    if (!project) return res.status(404).json({ detail: 'Project not found' });

    const key = await Counter.getNextKey(project_id, project.key);
    const { title, description = '', status = 'open', priority = 'medium', assignee_id = null } = req.body;
    const now = new Date().toISOString();
    const bugId = uuidv4();

    const doc = await Bug.create({
      id: bugId, project_id, key, title, description, status, priority,
      assignee_id, created_by: req.user._id, created_by_name: req.user.name,
      created_at: now, updated_at: now,
    });

    await Comment.create({
      id: uuidv4(), entity_id: bugId, entity_type: 'bug',
      user_id: req.user._id, user_name: req.user.name,
      content: 'reported this bug', type: 'activity', created_at: now,
    });

    if (assignee_id && assignee_id !== req.user._id) {
      await Notification.create({
        id: uuidv4(), user_id: assignee_id, type: 'bug_assigned',
        message: `Bug ${key}: ${title} assigned to you`,
        link: `/project/${project_id}/bugs`, created_at: now,
      });
    }

    const result = doc.toObject();
    delete result._id; delete result.__v;
    res.json(result);
  } catch (err) {
    res.status(500).json({ detail: err.message });
  }
});

// PUT /api/bugs/:id
router.put('/:id', auth, async (req, res) => {
  try {
    const old = await Bug.findOne({ id: req.params.id }).select('-_id -__v').lean();
    if (!old) return res.status(404).json({ detail: 'Bug not found' });

    const updateData = {};
    for (const [k, v] of Object.entries(req.body)) {
      if (v !== undefined && v !== null) updateData[k] = v;
    }
    updateData.updated_at = new Date().toISOString();
    await Bug.updateOne({ id: req.params.id }, { $set: updateData });
    const now = new Date().toISOString();

    if (updateData.status && updateData.status !== old.status) {
      await Comment.create({
        id: uuidv4(), entity_id: req.params.id, entity_type: 'bug',
        user_id: req.user._id, user_name: req.user.name,
        content: `changed status to ${updateData.status.replace(/_/g, ' ').replace(/\b\w/g, c => c.toUpperCase())}`,
        type: 'activity', created_at: now,
      });
      const target = old.assignee_id || old.created_by;
      if (target && target !== req.user._id) {
        await Notification.create({
          id: uuidv4(), user_id: target, type: 'status_changed',
          message: `${old.key} status changed to ${updateData.status}`,
          link: `/project/${old.project_id}/bugs`, created_at: now,
        });
      }
    }

    if (updateData.assignee_id && updateData.assignee_id !== old.assignee_id) {
      const assigneeUser = await User.findById(updateData.assignee_id);
      const assigneeName = assigneeUser?.name || 'someone';
      await Comment.create({
        id: uuidv4(), entity_id: req.params.id, entity_type: 'bug',
        user_id: req.user._id, user_name: req.user.name,
        content: `assigned this to ${assigneeName}`, type: 'activity', created_at: now,
      });
    }

    const updated = await Bug.findOne({ id: req.params.id }).select('-_id -__v').lean();
    res.json(updated);
  } catch (err) {
    res.status(500).json({ detail: err.message });
  }
});

// DELETE /api/bugs/:id
router.delete('/:id', auth, async (req, res) => {
  try {
    await Bug.deleteOne({ id: req.params.id });
    await Comment.deleteMany({ entity_id: req.params.id });
    res.json({ message: 'Bug deleted' });
  } catch (err) {
    res.status(500).json({ detail: err.message });
  }
});

// POST /api/bugs/:id/attachments
router.post('/:id/attachments', auth, async (req, res) => {
  try {
    const bug = await Bug.findOne({ id: req.params.id });
    if (!bug) return res.status(404).json({ detail: 'Bug not found' });
    const { file_name, file_url, file_type = 'image/png' } = req.body;
    const att = {
      id: uuidv4(), file_name, file_url, file_type,
      uploaded_by: req.user._id, created_at: new Date().toISOString(),
    };
    await Bug.updateOne({ id: req.params.id }, { $push: { attachments: att } });
    res.json(att);
  } catch (err) {
    res.status(500).json({ detail: err.message });
  }
});

// DELETE /api/bugs/:id/attachments/:attId
router.delete('/:id/attachments/:attId', auth, async (req, res) => {
  try {
    await Bug.updateOne({ id: req.params.id }, { $pull: { attachments: { id: req.params.attId } } });
    res.json({ message: 'Attachment removed' });
  } catch (err) {
    res.status(500).json({ detail: err.message });
  }
});

module.exports = router;
